File: src/org/jgll/recognizer/Descriptor.java
package org.jgll.recognizer;

import org.jgll.grammar.GrammarSlot;
import org.jgll.parser.HashFunctions;
import org.jgll.util.hashing.Decomposer;

/**
 * 
 * 
 * @author Ali Afroozeh
 *
 */
public class Descriptor {
	
	/**
	 * The label that indicates the parser code to execute for the encountered
	 * nonterminal.
	 */
	private final GrammarSlot slot;
	
	/**
	 * The associated GSSNode.
	 */
	private final GSSNode gssNode;
	
	/**
	 * The current index in the input string.
	 */
	private final int inputIndex;
	
	
	public Descriptor(GrammarSlot slot, GSSNode gssNode, int inputIndex) {
		
		assert slot != null;
		assert gssNode != null;
		assert inputIndex >= 0;
		
		this.slot = slot;
		this.gssNode = gssNode;
		this.inputIndex = inputIndex;
	}
	
	public GrammarSlot getGrammarSlot() {
		return slot;
	}

	public GSSNode getGSSNode() {
		return gssNode;
	}
	
	public int getInputIndex() {
		return inputIndex;
	}
	
	@Override
	public int hashCode() {
		return HashFunctions.defaulFunction().hash(slot.getId(), inputIndex, gssNode.hashCode());
	}
	
	@Override
	public boolean equals(Object obj) {
		if(this == obj) { 
			return true;
		}
		
		if(! (obj instanceof Descriptor)) {
			return false;
		}
		
		Descriptor other = (Descriptor) obj;

		return inputIndex == other.getInputIndex() &&
			   slot.equals(other.slot) &&
			   gssNode.equals(other.getGSSNode());
	}
	
	@Override
	public String toString() {
		return "(" + slot + ", " + inputIndex + ", " + gssNode + ")";
	}

	public static class DescriptorDecomposer implements Decomposer<Descriptor> {

		private int[] components = new int[4];
		
		@Override
		public int[] toIntArray(Descriptor descriptor) {
			components[0] = descriptor.slot.getId();
			components[1] = descriptor.inputIndex;
			components[2] = descriptor.gssNode.getGrammarSlot().getId();
			components[3] = descriptor.gssNode.getInputIndex();
			return components;
		}
	}
	
}

File: src-scala/org/jgll/recognizer/Descriptor.scala
package org.jgll.recognizer

import org.jgll.grammar.slot.GrammarSlot
import org.jgll.parser.HashFunctions
import org.jgll.util.hashing.ExternalHasher
import org.jgll.util.hashing.hashfunction.HashFunction
import Descriptor._
import scala.reflect.{BeanProperty, BooleanBeanProperty}
//remove if not needed
import scala.collection.JavaConversions._

object Descriptor {

  val externalHasher = new DescriptorExternalHasher()

  @SerialVersionUID(1L)
  class DescriptorExternalHasher extends ExternalHasher[Descriptor] {

    override def hash(descriptor: Descriptor, f: HashFunction): Int = {
      f.hash(descriptor.slot.getId, descriptor.inputIndex, descriptor.gssNode.getGrammarSlot.getId, descriptor.gssNode.getInputIndex)
    }

    override def equals(d1: Descriptor, d2: Descriptor): Boolean = {
      d1.inputIndex == d2.getInputIndex && d1.slot == d2.slot && 
        d1.gssNode.getGrammarSlot.getId == d2.gssNode.getGrammarSlot.getId && 
        d1.gssNode.getInputIndex == d2.gssNode.getInputIndex
    }
  }
}

class Descriptor(private val slot: GrammarSlot, private val gssNode: GSSNode, @BeanProperty val inputIndex: Int)
    {

  assert(slot != null)

  assert(gssNode != null)

  assert(inputIndex >= 0)

  def getGrammarSlot(): GrammarSlot = slot

  def getGSSNode(): GSSNode = gssNode

  override def hashCode(): Int = {
    externalHasher.hash(this, HashFunctions.defaulFunction())
  }

  override def equals(obj: Any): Boolean = {
    if (this == obj) {
      return true
    }
    if (!(obj.isInstanceOf[Descriptor])) {
      return false
    }
    val other = obj.asInstanceOf[Descriptor]
    inputIndex == other.getInputIndex && slot == other.slot && 
      gssNode.getGrammarSlot.getId == other.gssNode.getGrammarSlot.getId && 
      gssNode.getInputIndex == other.gssNode.getInputIndex
  }

  override def toString(): String = {
    "(" + slot + ", " + inputIndex + ", " + gssNode + ")"
  }
}
